Add IntegrityError handling back in as test expects.

doctoryes opened the pull request:
https://openedx.atlassian.net/browse/TNL-3692

@symbolist had taken out the IntegrityError handling in this code. But the failing test indicates that a recovery of sorts is expected in this case. So get the record and return it upon IntegrityError.

@symbolist Please review - transaction-related.

@macdiesel @nedbat @alawibaba @muzaffaryousaf @muhammad-ammar Please review.

macdiesel commented:
I'm cool with it if @symbolist is cool with it? :+1:

symbolist commented on lms/djangoapps/verify_student/models.py:
```diff
@@ -1112,8 +1112,13 @@ def get_or_create_verification_checkpoint(cls, course_id, checkpoint_location):
+                return checkpoint
+        except IntegrityError:
+            # Record already exists - just return it.
+            return cls.objects.get(course_id=course_id, checkpoint_location=checkpoint_location)
```
@doctoryes This is always going to raise a `DidNotExist`. The `IntegrityError` happens because in a parallel request a row in the table has been created, however in the `get_or_create`:
1. The `get` will not see that row because it was committed after the transaction in which this `get` is being executed was started.
2. So `create` will be called. However, MySQL will raise an `IntegrityError` because it sees that a parallel request has already created the row and this transaction cannot be committed later.

So doing a `get` will fail with a `DoesNotExist.` In Django 1.4, we had a `commit()` call in the `except` clause which started a new transaction which could see the row and therefore the `get` here succeeded.

doctoryes replied:
I agree - thanks for catching this.

A couple of options here:
* Change this to return `None` when get_or_create fails and let the higher-level code decide what to do (probably re-try). Also change the test accordingly. 
* Re-try within this method to send back a proper record, which some parts of the code expects.

symbolist replied:
Since this method is called from an XBlock service, I doubt we can use `outer_atomic` here. So if an `IntegrityError` is raised we cannot do anything about it. The only thing I can think of doing here is using `READ COMMITTED` for the view which is calling this method. `courseware.views.index` is already using `READ COMMITTED` so that would make an `IntegrityError` here very unlikely. If it does happen I would let it propogate upwards for now.

doctoryes commented:
@symbolist I removed the IntegrityError handling altogether, changed the test, and added a comment about what views that call this method should do. Re-review?

macdiesel commented on lms/djangoapps/verify_student/models.py:
```diff
@@ -1112,8 +1112,9 @@ def get_or_create_verification_checkpoint(cls, course_id, checkpoint_location):
         Returns:
             VerificationCheckpoint object if exists otherwise None
```
We should a Raises: to the docs so callers will know what exceptions to expect.

doctoryes replied:
@macdiesel Added.

symbolist commented:
:+1:

macdiesel commented:
Merge it! :+1:
